### PR TITLE
Perfect fruit shop

### PR DESCRIPTION
**Pull request by StanislavSonich**

### CODE REVIEW

**Comment by StanislavSonich** [comment]
Fruin shop is opening now.

File: src/main/java/dao/FruitDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        return fruit;
+    }
+
+    @Override
+    public Fruit get(String name) {
+        return Storage.fruits.stream()
+                .filter(f -> f.getName().equals(name))
+                .findFirst()
+                .get();
+    }
```
Comment (Rommelua) [inline src/main/java/dao/FruitDaoImpl.java:20]: Make Storage as Map<String, Integer> and you will have O(1) complexity for get

File: src/main/java/model/Fruit.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package model;
+
+public class Fruit {
```
Comment (Rommelua) [inline src/main/java/model/Fruit.java:3]: You can delete this class. **quantity** is not a Fruit property. quantity is Storage property.

File: src/main/java/service/CreateReport.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package service;
+
+import java.util.List;
+import model.Fruit;
+
+public interface CreateReport {
+    String createReport(List<Fruit> fruits);
```
Comment (Rommelua) [inline src/main/java/service/CreateReport.java:7]:
```suggestion
    String createReport();
```
go to Storage inside implementation

File: src/main/java/service/FruitService.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package service;
+
+import java.util.List;
+import model.Fruit;
+import model.TransactionDto;
+
+public interface FruitService {
+    List<Fruit> changeBalance(List<TransactionDto> transactions);
```
Comment (Rommelua) [inline src/main/java/service/FruitService.java:8]:
```suggestion
    void executeTransactions(List<Transaction> transactions);
```

File: src/main/java/service/Parser.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package service;
+
+import model.TransactionDto;
+
+public interface Parser {
+    TransactionDto parseLine(String line);
```
Comment (Rommelua) [inline src/main/java/service/Parser.java:6]:
```suggestion
    List<Transaction> parse(List<String> lines);
```
Let`s remove the loop from main() method

File: src/main/java/model/TransactionDto.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package model;
+
+public class TransactionDto {
```
Comment (Rommelua) [inline src/main/java/model/TransactionDto.java:3]: image

File: src/main/java/service/impl/FileReaderImpl.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import service.ReaderService;
+
+public class FileReaderImpl implements ReaderService {
+    public List<String> readFromFile(String filePath) {
+        File file = new File(filePath);
+        if (!file.exists()) {
+            throw new RuntimeException("Can't find file by path: " + filePath);
+        }
```
Comment (Rommelua) [inline src/main/java/service/impl/FileReaderImpl.java:16]:
```suggestion
```

File: src/main/java/service/impl/FileReaderImpl.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        try (BufferedReader reader = new BufferedReader(new FileReader(file.getPath()))) {
+            List<String> text = new ArrayList<>();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                text.add(line);
+            }
+            return text;
+        } catch (IOException e) {
+            throw new RuntimeException("Can't read file!", e);
```
Comment (Rommelua) [inline src/main/java/service/impl/FileReaderImpl.java:25]:
```suggestion
            throw new RuntimeException("Can't read file " + filePath, e);
```

File: src/main/java/service/impl/FileWriterImpl.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import service.FileWriterService;
+
+public class FileWriterImpl implements FileWriterService {
+    public boolean writeData(String toFilePath, String data) {
+        File file = new File(toFilePath);
+        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file, false))) {
+            bufferedWriter.write(data);
+            return true;
+        } catch (IOException e) {
+            throw new RuntimeException("Can't write data to file!", e);
```
Comment (Rommelua) [inline src/main/java/service/impl/FileWriterImpl.java:16]: image

File: src/main/java/service/impl/FileWriterImpl.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import service.FileWriterService;
+
+public class FileWriterImpl implements FileWriterService {
+    public boolean writeData(String toFilePath, String data) {
+        File file = new File(toFilePath);
+        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file, false))) {
```
Comment (Rommelua) [inline src/main/java/service/impl/FileWriterImpl.java:12]:
```suggestion
        try (BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(file))) {
```

File: src/main/java/service/impl/ValidatorServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+    public boolean validate(List<String> text) {
+        if (text.isEmpty() || !text.get(0).equals(TITLE_TEXT)) {
+            throw new RuntimeException(("Incorrect input data!"));
+        }
+        for (int i = 1; i < text.size(); i++) {
+            String line = text.get(i);
+
+            if (!line.matches(FORMAT_TEXT)) {
+                throw new RuntimeException("Incorrect format data!");
```
Comment (Rommelua) [inline src/main/java/service/impl/ValidatorServiceImpl.java:17]: make exception msgs more informative

File: src/main/java/strategy/Operation.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        return operationTypes;
+    }
+
+    public static Operation getActivityType(String abbreviationActivityType) {
+        for (Operation typeActivity: Operation.values()) {
+            if (typeActivity.getOperationTypes().equals(abbreviationActivityType)) {
+                return typeActivity;
+            }
+        }
+        return null;
```
Comment (Rommelua) [inline src/main/java/strategy/Operation.java:25]: do not return null - throw exception

File: src/main/java/strategy/OperationHandler.java [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package strategy;
+
+public interface OperationHandler {
+    void apply(String name, int quantity);
```
Comment (Rommelua) [inline src/main/java/strategy/OperationHandler.java:4]:
```suggestion
    void handle(Transaction transaction);
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Pull request closed without merging**